lib/ClangImporter/ImportDecl.cpp

```diff
@@ -4322,17 +4316,50 @@ namespace {
     T *resolveSwiftDecl(const U *decl, Identifier name,
                         bool hasKnownSwiftName, ClangModuleUnit *clangModule) {
       if (auto overlay = clangModule->getOverlayModule())
-        return resolveSwiftDeclImpl<T>(decl, name, hasKnownSwiftName, overlay);
+        return resolveSwiftDeclImpl<T>(decl, name, hasKnownSwiftName, overlay,
+                                       /*allowObjCMismatchFallback*/ true);
       if (clangModule == Impl.ImportedHeaderUnit) {
         // Use an index-based loop because new owners can come in as we're
         // iterating.
         for (size_t i = 0; i < Impl.ImportedHeaderOwners.size(); ++i) {
           ModuleDecl *owner = Impl.ImportedHeaderOwners[i];
-          if (T *result = resolveSwiftDeclImpl<T>(decl, name,
-                                                  hasKnownSwiftName, owner))
+          if (T *result =
+                  resolveSwiftDeclImpl<T>(decl, name, hasKnownSwiftName, owner,
+                                          /*allowObjCMismatchFallback*/ true))
             return result;
         }
       }
+      if (Impl.ImportForwardDeclarations) {
```

Comment:
I think I'm missing something important here: *Why* are we doing this? I can see that we're looking in the use sites imported modules for the name we're trying to import, and that we're only doing it when `ImportForwardDeclarations` is enabled, but I don't understand the overall purpose. Why do we need to look in the imported modules?

I'm basically asking because looking in the imported modules of the use sites seems *really, really weird* and I'm not sure it will interact well with the way ClangImporter caches declarations. (After all, couldn't we try to import the same declaration from use sites with different imported modules? What happens then?)

Reply:
So this `resolveSwiftDecl` function, even prior to this patch was used only to resolved forward declared interfaces and protocols in Objective-C headers to a Swift definition. As I understand it, before this patch, it only looked in the overlay module to support mixed modules.

**Why are we now also looking at the site's imported modules?**

Well to start, to match the behavior of Clang modules. When you import Clang modules A and B into a Swift file, a forward declaration `@class Foo` in module A, will be resolved to a definition `@interface Foo ...` in module B. When you use the same `clang::CompilerInstance` to load two modules, the AST is "merged". But if the definition is instead in **Swift** module C, this won't work anymore.

This is really confusing for users not versed in Clang modules vs Swift modules. Perhaps misguided, but the conventional advice for dealing with forward declarations without bloating your ObjC headers with unnecessary includes has been to "import the defining module". 

Specifically with reference to this patch, not doing this leaves us vulnerable to other problems. You can see an example in `incomplete_objc_types_swift_definition_imported.swift`. If we start importing placeholder types for forward declarations, _without_ searching for a visible Swift @objc definition, we can end up with two conflicting Swift representations for the same type. 

**Why are we only doing it when ImportForwardDeclarations is enabled?**

As mentioned, it is required for importing forward declarations safely and source breaking so needs to be gated behind the flag. It would no doubt be useful separate from this feature, but that seems like a different patch.

**Isn't it an issue if different use sites have different imported modules?**

I don't think so. We are only looking to resolve the name to an @objc annotated Swift declaration, nothing else. To my knowledge, @objc provided names are not "namespaced" by the defining module. I don't think `@objc(Foo) class Bar {}` and `@objc(Foo) class Baz {}` in the same program is supported, regardless of which module they are defined in (I've seen warnings along the lines of "multiple selectors for Foo, which implementation is used is undefined"). All that to say if we find one definition, that should be the only one. Similarly once one definition 
 is cached, it should be fine to continue returning that. That said, I could be wrong about the rules around @objc name conflicts, so let me know if I've missed something.

Reply:
EDIT: I've uncovered problems, need to take another run at it.

Reply:
> If we start importing placeholder types for forward declarations, without searching for a visible Swift @objc definition first, we can end up with two conflicting Swift representations for the same type (one directly from the imported Swift module, the other created to represent a forward declaration of the same Swift type).

To be clear, if we *do* find a type this way, is the API using it still imported as unavailable? Or does it mean that APIs that were previously dropped because they relied on a non-modular forward declaration are now imported *and usable*? The latter doesn't seem to be within the scope of the Evolution proposal, and I have concerns about whether it might end up producing corrupt module files (because some of their declarations use types from modules that aren't in the module's import list).

> If it helps, we could also look at all Swift modules imported into the entire main module? I don't really like that though.

I think I would prefer that because it would mean ClangImporter was always considering the same set of imported modules, at least within a given compilation job.

Reply:
> To be clear, if we do find a type this way, is the API using it still imported as unavailable? Or does it mean that APIs that were previously dropped because they relied on a non-modular forward declaration are now imported and usable? 

I'm not sure what you mean by a non-modular forward declaration, but yes the API relying on it would then be imported and usable. 

> The latter doesn't seem to be within the scope of the Evolution proposal, and I have concerns about whether it might end up producing corrupt module files (because some of their declarations use types from modules that aren't in the module's import list).

Fair enough, admittedly this problem didn't occur to me until later in the evolution process. If that's the case, I will have to start another round of evolution as I don't think this change is feasible without this.

Could you elaborate a little on the module's import list? I'm not familiar. In what scenario would a module be available while compiling a module, but not included in the produced module's import list?

> I think I would prefer that because it would mean ClangImporter was always considering the same set of imported modules, at least within a given compilation job.

Yeah I've made this change, I don't think we have any other choice. Could you elaborate on the compilation job comment? I was investigating a bit if the imported modules might change between each front-end invocation of an incremental build, but in the testing I did, it seemed ok? Am I wrong in this belief?